File: 2024/rs/src/bin/day02.rs
fn main() {
    let input = aoc::read_stdin().expect("cannot read stdin");
    // println!("{input}");

    let data = parse(&input).expect("cannot parse input");
    // println!("{data:?}");

    let result1 = part1(&data);
    println!("part1: {result1}");

    let result2 = part2(&data);
    println!("part2: {result2}");
}

type Data = Vec<Vec<usize>>;

fn parse(input: &str) -> Result<Data, String> {
    Ok(input
        .lines()
        .map(|vs| {
            vs.split_whitespace()
                .map(|v| v.parse::<usize>().expect("cannot parse as usize"))
                .collect()
        })
        .collect())
}

fn part1(data: &Data) -> usize {
    data.iter()
        .map(|vs| is_safe_seq(vs))
        .filter(|v| *v == true)
        .count()
}

fn part2(data: &Data) -> usize {
    data.iter()
        .map(|vs| {
            // Produce all permutations of vs with one element removed
            let vsps: Vec<_> = (0..vs.len())
                .into_iter()
                .map(|i| {
                    let mut vsp = vs.clone();
                    vsp.remove(i);

                    vsp
                })
                .collect();

            // Yield if any of the permuations is safe
            vsps.iter().any(|vs| is_safe_seq(vs))
        })
        .filter(|v| *v == true)
        .count()
}

fn is_safe_seq(vs: &[usize]) -> bool {
    // Produce a list of differences between adjacent values
    let vsp: Vec<_> = vs
        .windows(2)
        .map(|w| w[1] as isize - w[0] as isize)
        .collect();

    let is_all_increasing = |vs: &[isize]| -> bool { vs.iter().all(|v| *v > 0) };
    let is_all_decreasing = |vs: &[isize]| -> bool { vs.iter().all(|v| *v < 0) };
    let is_all_diff_gte_1 = |vs: &[isize]| -> bool { vs.iter().all(|v| v.abs() >= 1) };
    let is_all_diff_lte_3 = |vs: &[isize]| -> bool { vs.iter().all(|v| v.abs() <= 3) };

    // Yield if the list of differences satisfies the safe criteria
    (is_all_increasing(&vsp) || is_all_decreasing(&vsp))
        && is_all_diff_gte_1(&vsp)
        && is_all_diff_lte_3(&vsp)
}

File: 2024/hs/src/day02.hs
import Data.List
import qualified Data.Map as M
import qualified Data.Maybe

main = do
    input <- getContents -- aka read_stdin
    -- putStrLn ("input: " ++ input)
    case parse input of
        Left err -> do
            putStrLn ("failed parsing input: " ++ err)
        Right v -> do
            -- putStrLn ("data: " ++ show v)
            putStrLn ("part1: " ++ show (part1 v))
            putStrLn ("part2: " ++ show (part2 v))

parse :: String -> Either String Data
parse = Right . map (map toInt . words) . lines
  where
    toInt x = read x :: Int

type Data = [[Int]]

part1 :: Data -> Maybe Int
part1 = Just . length . filter id . map isSafeSeq

part2 :: Data -> Maybe Int
part2 = Just . length . filter id . map (any isSafeSeq . expand)
  where
    expand :: [a] -> [[a]]
    expand xs = zipWith removeAt [0 ..] $ replicate n xs
      where
        n = length xs
        removeAt i xs = take i xs ++ drop (i + 1) xs

isSafeSeq :: (Num a, Ord a) => [a] -> Bool
isSafeSeq xs = (allIncreasing || allDecreasing) && allAbsGte1 && allAbsLte3
  where
    vs = zipWith (-) xs (tail xs)
    allIncreasing = all (> 0) vs
    allDecreasing = all (< 0) vs
    allAbsGte1 = all ((>= 1) . abs) vs
    allAbsLte3 = all ((<= 3) . abs) vs
